make python rpc handler to be singleton class

zhaojuanmao opened the pull request:
Summary: python rpc handler right now is namespace + global variable, changed it to be singleton class as it can gurantee deterministic order of variable destruction. for namespace + global variable, we hit a process exit crash issue because global variables have dependencies and they are not destructed as expected

Test Plan: unit test passed

Differential Revision: D17097999

xush6528 commented on torch/csrc/distributed/rpc/python_rpc_handler.h:
```diff
@@ -7,17 +7,34 @@ namespace torch {
+  PythonRpcHandler(PythonRpcHandler&&) = delete;
+  PythonRpcHandler& operator=(PythonRpcHandler&&) = delete;
+
+  py::object runUDFFunction_;
```
```
py::object runUDFFunction_ = py::none();
```

pritamdamania87 replied:
Initialization should be done in the constructor, not here.

xush6528 commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -3,43 +3,41 @@
+  AutoGIL ag;
+  py::object module =
+      py::module::import("torch.distributed.internal_rpc_utils");
+  if (runUDFFunction_ == nullptr) {
```
```
runUDFFunction_.is(py::none())
```

pritamdamania87 replied:
We don't need these if statements in the constructor since this will happen only once.

xush6528 reviewed (commented):
Better way

pietern commented:
As is, this will conflict with #25558.

Can you make the singleton change in a PR separate from that one?

zhaojuanmao committed:
make rpc multiprocess test to use spawn instead of fork

Summary:
1. current fork unit tests did not catch rpc exit crash issue, but spawn caught it. pyper also uses spawn,
so changing multiprocessing start method from spawn to fork

2. it did not change fork to spawn for c10d tests for now -- that needs some discussion, any reason we do not want c10d to use spawn?

3. also fixed the bug that is being fixed in https://github.com/pytorch/pytorch/pull/25144, will rebase once fix is landed

4. fix the real segment exit issue caused by python rpc handler temporarily by removing module_ global variable, will send a followup diff to change python rpc handler to be singletion class for deterministic destruction order of variables

Test Plan:
use fork, rpc tests passed;
use spawn, rpc tests crashed while exiting; rpc tests passed after fixing python rpc handler

Differential Revision: D17086007

fbshipit-source-id: 0284bf3d7d0e241bbe258df1c0273ac285b1988a

zhaojuanmao committed:
make python rpc handler to be singleton class (#25559)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/25559

python rpc handler right now is namespace + global variable, changed it to be singleton class as it can gurantee deterministic order of variable destruction. for namespace + global variable, we hit a process exit crash issue because global variables have dependencies and they are not destructed as expected

Test Plan: unit test passed

Differential Revision: D17097999

fbshipit-source-id: 19ebfee0e0c2c067dc8555c3da544ed691bf4387

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.h:
```diff
@@ -7,17 +7,34 @@ namespace torch {
-}
+// singleton class provides interface to  execute python UDF remote call
+// and deserialize the returned results by running python function
+// in internal_rpc_utilities
```
do you mean `torch/distributed/internal_rpc_utils.py`?

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.h:
```diff
@@ -7,17 +7,34 @@ namespace torch {
+ public:
+  static PythonRpcHandler& getInstance();
+  // execute python UDF, result is pickled to binary string
+  std::vector<char> generatePythonUDFResult(const Message& request);
```
Why these two methods are not static?

mrshenli replied:
Does it work if we make all methods static instead of using a singleton class?

mrshenli replied:
nvm, I realized that  `runUDFFunction_` and `loadResultFunction_ ` are now instance variables.

pritamdamania87 commented on test/common_distributed.py:
```diff
@@ -129,26 +129,39 @@ def setUp(self):
         self.rank = self.MAIN_PROCESS_RANK
-        self.file = tempfile.NamedTemporaryFile(delete=False)
+        # spawn can not pass file handle, so pass file name here
+        self.file = tempfile.NamedTemporaryFile(delete=False).name
```
Looks like this PR has changes from the 'spawn' PR. Can we export using `ghexport` to avoid this?

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -119,7 +119,8 @@ ProcessGroupAgent::ProcessGroupAgent(
 
-  PythonRpcHandler::init();
+  // construct PythonRpcHandler singleton here
+  PythonRpcHandler::getInstance();
```
We don't need to call this explicitly. Whenever the first `getInstance` method is encountered, the singleton would be created automatically.

zhaojuanmao replied:
will it be better to explicitly create it in main thread? if multiple worker threads try to create it, they have to wait the lock

pritamdamania87 commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -3,43 +3,41 @@
-      loadResultFunction_ = module_.attr("load_python_udf_result_internal");
-    }
+PythonRpcHandler::PythonRpcHandler() {
+  AutoGIL ag;
```
Do we need GIL even for importing the functions?

zhaojuanmao replied:
not just importing, but also write to the instance variables, so it needs GIL?

pritamdamania87 commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -3,43 +3,41 @@
+  AutoGIL ag;
+  auto pargs = py::bytes(request.payload().data(), request.payload().size());
+  TORCH_CHECK(runUDFFunction_ != nullptr, "runUDFFunction_ is nullptr");
+  py::bytes pres = runUDFFunction_(pargs);
```
IIUC, only `runUDFFunction_` needs to be protected by GIL? Can we move the rest of the code outside GIL?

pritamdamania87 commented on torch/csrc/distributed/rpc/python_rpc_handler.h:
```diff
@@ -7,17 +7,34 @@ namespace torch {
-// function to unpickle the python UDF result and return pyObject to user
-py::object loadPythonUDFResult(const Message& message);
-}
+// singleton class provides interface to  execute python UDF remote call
```
Nit: start with caps.

pritamdamania87 commented on torch/csrc/distributed/rpc/python_rpc_handler.h:
```diff
@@ -7,17 +7,34 @@ namespace torch {
+// singleton class provides interface to  execute python UDF remote call
+// and deserialize the returned results by running python function
+// in internal_rpc_utilities
+// the singleton object is constructed at first when RPC agent is
```
Nit: Start with caps.

pritamdamania87 commented on torch/csrc/distributed/rpc/python_rpc_handler.h:
```diff
@@ -7,17 +7,34 @@ namespace torch {
+  PythonRpcHandler(PythonRpcHandler&&) = delete;
+  PythonRpcHandler& operator=(PythonRpcHandler&&) = delete;
+
+  py::object runUDFFunction_;
```
Make both of these const and only initialize in the constructor.

zhaojuanmao replied:
tried const, but it is not allowed to assign py::object to const py::object when initialize

zhaojuanmao commented:
closing this one, #25742 by ghexport

pritamdamania87 commented:
@zhaojuanmao Shouldn't we land PRs only after an approval?

zhaojuanmao commented:
it is not landed, just closed. 

https://github.com/pytorch/pytorch/pull/25742 is the one landed after approval

zhaojuanmao commented:
I requested review from you on #25742

pritamdamania87 commented:
@zhaojuanmao Sorry, my bad I misread your comment.
